### ISSUE TITLE
[BUG] Has deepspeed completed activation value recalculation early to reduce latency?

### ISSUE THREAD

**Issue by Echozqn** [microsoft/DeepSpeed#4736]
**Describe the bug**
I saw that Gpipe used to recalculate the activation value of the next stage in advance to reduce delays, but when I used deepspeed training myself, I drew a timeline diagram of each stage. The backward time is about 3 times the forward time and I found that the early recomputation strategy is not used.

In the figure, purple represents backward and green represents forward. The depth of the pipeline is 3, and the gradient accumulation is 4.
<img width="414" alt="image" src="https://github.com/microsoft/DeepSpeed/assets/50656301/5c793c6b-d0f2-4905-95d8-1237ceb7c19e">

The picture below is the early recomputation strategy of Gpipe that I saw in the paper Merak.
<img width="557" alt="image" src="https://github.com/microsoft/DeepSpeed/assets/50656301/19912859-50df-443f-bfb5-25f8983f2b4d">
